src/loaders/genome_collection/parse_tool_results.py

```diff
@@ -475,247 +401,37 @@ def _create_meta_lookup(root_dir, env, kbase_collection, load_ver, tool):
-
+def _build_heatmap_meta(
+        reference_meta: list[dict],
+        kbase_collection: str,
+        load_ver: str,
+        min_value: float,
+        max_value: float) -> dict:
+    # Build the heatmap metadata from the reference metadata (list of metadata)
 
-def _is_binary(num: int) -> bool:
-    # Given a number, checks if it is a binary num (i.e., being only 0 or 1).
-    return num == 0 or num == 1
+    heatmap_categories = dict()
+    for meta in reference_meta:
+        category = meta[FIELD_HEATMAP_CATEGORY]
+        if category not in heatmap_categories:
+            heatmap_categories[category] = {FIELD_HEATMAP_CATEGORY: category, FIELD_HEATMAP_COLUMNS: []}
 
+        meta_without_category = meta.copy()
+        del meta_without_category[FIELD_HEATMAP_CATEGORY]
+        heatmap_categories[category][FIELD_HEATMAP_COLUMNS].append(meta_without_category)
 
-def _num_to_bool(num: int | bool) -> bool:
-    # Given a number, checks if it is a binary type (i.e., contains only 0s and 1s).
+    # sort the columns by column id
+    for category in heatmap_categories.values():
+        category[FIELD_HEATMAP_COLUMNS] = sorted(category[FIELD_HEATMAP_COLUMNS],
+                                                 key=lambda x: int(x[FIELD_HEATMAP_COL_ID]))
 
-    if _is_binary(num):
-        return bool(num)
-    else:
-        raise ValueError(f'Input must be a binary number (i.e., 0 or 1). Got {num} instead.')
+    heatmap_meta = {FIELD_HEATMAP_CATEGORIES: list(heatmap_categories.values()),
+                    FIELD_HEATMAP_MIN_VALUE: min_value,
+                    FIELD_HEATMAP_MAX_VALUE: max_value,
+                    names.FLD_ARANGO_KEY: collection_load_version_key(kbase_collection, load_ver),
+                    names.FLD_COLLECTION_ID: kbase_collection,
+                    names.FLD_LOAD_VERSION: load_ver}
 
-
-def _create_cell_detail(
-        cell_id: str,
-        detected_genes_score: dict[str, float]
-) -> CellDetail:
-    if not detected_genes_score:
-        detected_genes_score = dict()
-
-    return CellDetail(
-        cell_id=cell_id,
-        values=[
-            CellDetailEntry(id=gene_name, val=gene_score)
-            for gene_name, gene_score in detected_genes_score.items()
-        ]
-    )
-
-
-def _append_cell(
-        heatmap_row: HeatMapRow,
-        trait_idx: int,
-        cell_count: int,
-        trait_type: str,
-        min_value: float | int,
-        max_value: float | int,
-        trait_val: float | int | bool = _SYS_DEFAULT_TRAIT_VALUE,
-        detected_genes_score: dict[str, float] = None,
-) -> (float | int, float | int, CellDetail):
-    # Append a cell to the heatmap row and return the global min and max values
-
-    if trait_type == 'count':
-        trait_val = _int(trait_val)
-    elif trait_type == 'binary':
-        trait_val = _num_to_bool(trait_val)
-    else:
-        raise ValueError(f'Unknown trait type {trait_type}')
```

Comment:
Does this conversion need to be done in the tool container? I don't remember seeing anything like this

Reply:
👍 added.

Reply:
:+1: